rfcs/0133-git-hashing.md

```diff
@@ -0,0 +1,346 @@
+
+IPFS also supports git hashing, and so we also provide a good way for people and institutions to "pin" the sources they need, especially if those sources include private ones Software Heritage won't have.
+Finally, per [this IPFS plugin](https://blog.obsidian.systems/software-heritage-bridge/), we can have a 3 way integration between IPFS, Software Heritage, and Nix.
+Data can be directly downloaded from Software Heritage via HTTPS, or indirectly via IPFS.
+
+This would hopefully supercede the direct Software Heritage store outlined above.
+This is because Software Heritage is primarily an archive, and we should be careful not to waste their limitted resources with copious egress bandwidth.
+IPFS can act as a CDN to not put as much load on Software Heritage's servers.
+
+### Not just IPFS
+
+Many of the IPFS-specific logic could in fact live in a plugin if this is desired.
+However, we still need to adjust core abstractions of Nix store layer (as described below) to interface with IPFS in the best possible way.
+Those same adjustments would allow Nix to work better with *any* content-addressing system, so alternatives networks/projects to IPFS can also be just as easily experimented with.
+
+As always with my work, the mantra (from Scheme) to follow is
+
+> *x* should be designed not by piling feature on top of feature, but by removing the weaknesses and restrictions that make additional features appear necessary.
+
+A ton of misc features have been added to Nix since 2.0, and we are very careful to not increase total ad-hoc complexity more than necessary.
+
+## Detailed Design
+
+Each item can be done separately provided its dependent items are also done.
+These are the items we do *not* wish to commit to at this time, but leave for later.
+(See [Detailed Design](#detailed-design) for other steps left for later.)
+
+### Augmented `narinfo`
+
+- **Purpose**: Binary distribution
+
+*This is taken from [RFC PR 122](https://github.com/lucasew/rfcs/blob/binary-cache-ipfs/rfcs/0122-binary-cache-ipfs.md), which was abandoned by its author.*
+
+The purpose of this is a "hybrid" store where the narinfo metadata is still served via HTTPS, but the data itself is served via IPFS.
+
+Today, a narinfo looks like this:
+
+```
+StorePath: /nix/store/gdh8165b7rg4y53v64chjys7mbbw89f9-hello-2.10
+URL: nar/0i6ardx43rdg24ab1nc3mq7f5ykyiamymh1v37gxdv5xh5cm0cmb.nar.xz
+Compression: xz
+FileHash: sha256:0i6ardx43rdg24ab1nc3mq7f5ykyiamymh1v37gxdv5xh5cm0cmb
+FileSize: 40360
+NarHash: sha256:1ddv0iqq47j0awyw7a8dmm8bz71c6ifrliq53kmmsfzjxf3rwvb8
+NarSize: 197528
+References: 7gx4kiv5m0i7d7qkixq2cwzbr10lvxwc-glibc-2.27 gdh8165b7rg4y53v64chjys7mbbw89f9-hello-2.10
+Deriver: 5sj6fdfym58sdaf3r5p87v4l8sj2zlvn-hello-2.10.drv
+Sig: cache.nixos.org-1:K0thQEG60rzAK8ZS9f1whb7eRlIshlMDJAm7xvX1oF284H+PTqlicv/wGW6BIj+wWWONHvUZ2MYc+KDArekjDA==
+```
+
+This RFC proposes new key-value pairs that in this example would be:
+
+```
+IpfsCid: Qmf8NfV2hnq44RoQw9vxmSpGYTwAovA8FUCxeCJCqmXeNN
```

Comment:
It's not entirely clear to me why we're using IPFS CIDs (and unixfs) here, instad of let's say, pointing to a raw git (blob or tree) object. This feels like a more logical choice, considering the first part starts with git tree hashing.

Can you make it more clear why unixfs was chosen?

In addition to that, I'd personally be in favor having `IpfsCid / IpfsEncoding` be one single field, and having specified in that RFC already how additional representations are handled. (Can these fields be repeated, or is this a ?-separated string of multiple possible representations?)

This allows us gradually introducing more representations, and pointing to more than one representation, without worrying about older Nix versions not supporting these protocols to not be able to parse it at all).

Reply:
>  I'd personally be in favor having IpfsCid / IpfsEncoding be one single field

I think it is best to keep it separated. For the common operation of retrieving the file you only need the CID. The encoding settings are only needed if you have the store path and want to add it to IPFS to share with others which is probably a relatively rare operation.

Reply:
Actually on further review it is somewhat mixed. `"method":"wrapped-nar"` is needed for both reading and resharing. Maybe it makes sense to move that property to `IpfsCid` (which will then need to be renamed) so that all reading properties are in the same place.

Reply:
As written above, we might need to specify how multiple different encodings are specified, and thinking about "describing the NAR file twice, but with different chunking parameters" would be one usecase where it'd be quite complicated if cid and "generation method/interpretation" were two separate fields.

Reply:
Would we ever want to describe the same NAR file twice with different chunking parameters? It seems that in general the creator should pick the parameters and to ensure maximum sharing everyone should re-use them.

In theory we don't ever need chunking parameters. Ideally the producer would share on IPFS and everyone would just download that CID. The main reason that we provide them is to help bootstrapping. For example right now https://cache.nixos.org isn't going to publish via IPFS, so the standardized chunking parameters will allow the community to do that. If the producer does publish the CID themselves then everyone would just download that and now worry about how it was chunked.

However interpretation is important as once you download the CID you need to know how to use it. It is needed for both reading and writing.

I agree that **if** we want to support multiple CIDs then it probably makes sense to group everything together to avoid parallel corresponding lists. However if we don't want to support mutliple CIDs then I think it makes sense to separate "things required to download and use" from "things required if you want to start seeding the CID".

Reply:
I think "This RFC" is misleading, because this is the the non-normative future work section, no?